### PR TITLE
Adds tools to easily paste Markdown from the clipboard (Resolves #1590)

### LINKED ISSUE #1590: Markdown doesn't automatically render on pasting content or on existing line
When you type the syntax first, the markdown is instantly rendered but it's not working in case of pasting text or on an existing line.

For example if I start with ## and then type `Some heading` it works but if I directly paste `## Some heading`, it doesn't work.

Similarly for lists, if start with `1.` then it works, but if I type `This is a line` and then type `1.` at the start, it doesn't work.

### PR DESCRIPTION
**Pull request by matthew-carroll**
Adds tools to easily paste Markdown from the clipboard (Resolves #1590)

To add Markdown pasting capability to a `SuperEditor` in an app, do the following:

```dart
SuperEditor(
  //...
  keyboardActions: [
    pasteMarkdownOnCmdAndCtrlV,
    ...defaultKeyboardActions,
  ],
 //...
);
```

This PR also...

  * Migrates `super_editor_markdown` to Dart 3.x.
  * Updates and fixes analysis warnings. 
  * Adds a new request/command pair to the default list, which can insert/paste structured content.

### CODE REVIEW

File: super_editor/lib/src/default_editor/multi_node_editing.dart [hunk 1]
```diff
@@ -14,6 +14,241 @@ import 'paragraph.dart';
...
+    required this.content,
+    required this.pastePosition,
+  });
+
+  final List<DocumentNode> content;
+  final DocumentPosition pastePosition;
+}
+
+/// Inserts the given structured [_content] in the document at the given [_pastePosition].
+class PasteStructuredContentEditorCommand implements EditCommand {
```
Comment (angelosilvestre) [inline super_editor/lib/src/default_editor/multi_node_editing.dart:30]: Shouldn't we declare `_content` and `_pastePosition` as public, since we are referencing them on the dart doc?

File: super_editor/lib/src/default_editor/multi_node_editing.dart [hunk 2]
```diff
@@ -14,6 +14,241 @@ import 'paragraph.dart';
...
+    }
+    if (currentNodeWithSelection.text.length == 0) {
+      // The node with the selection is an empty text node. After we use that node's
+      // position to insert other nodes, we want to delete that first node, as if the
+      // pasted content replaced it.
+      deleteInitiallySelectedNode = true;
+    }
+
+    // (Possibly) merge or the downstream split node.
+    final lastPastedNode = nodesToInsert.last;
```
Comment (angelosilvestre) [inline super_editor/lib/src/default_editor/multi_node_editing.dart:156]: Should we check if `nodesToInsert` isn't empty after removing the first node? Even though this method is called `_pasteMultipleNodes` we might want to handle it correctly if it's called with a single node on the list.

File: super_editor/lib/src/default_editor/multi_node_editing.dart [hunk 3]
```diff
@@ -14,6 +14,241 @@ import 'paragraph.dart';
...
+      // node.
+      executor.executeCommand(
+        InsertAttributedTextCommand(
+          documentPosition: DocumentPosition(
+            nodeId: downstreamSplitNode.id,
+            nodePosition: const TextNodePosition(offset: 0),
+          ),
+          // Only text nodes are merge-able, therefore we know that the first pasted node
+          // is a TextNode.
+          textToInsert: (lastPastedNode as TextNode).text,
```
Comment (angelosilvestre) [inline super_editor/lib/src/default_editor/multi_node_editing.dart:170]: "first" -> "last"

File: super_editor/lib/src/default_editor/multi_node_editing.dart [hunk 4]
```diff
@@ -14,6 +14,241 @@ import 'paragraph.dart';
...
+        ),
+      );
+
+      // We've pasted the first new node. Remove it from the nodes to insert.
+      nodesToInsert.removeLast();
```
Comment (angelosilvestre) [inline super_editor/lib/src/default_editor/multi_node_editing.dart:175]: "first" -> "last"

File: super_editor/lib/src/default_editor/multi_node_editing.dart [hunk 5]
```diff
@@ -14,6 +14,241 @@ import 'paragraph.dart';
...
+        SelectionChangeType.insertContent,
+        SelectionReason.userInteraction,
+      ),
+    );
+  }
+
+  (String upstreamNode, String downstreamNode) _splitPasteParagraph(
+    CommandExecutor executor,
+    String currentNodeWithSelectionId,
+    int pasteTextOffset,
```
Comment (angelosilvestre) [inline super_editor/lib/src/default_editor/multi_node_editing.dart:222]: Should we name the return values `upstreamNodeId` and `downstreamNodeId`?

File: super_editor_markdown/test/super_editor_markdown_pasting_test.dart [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+        ),
+      );
+
+      // Place the caret in the empty document.
+      await tester.placeCaretInParagraph("1", 0);
+
+      // Ensure that the document has the caret.
+      expect(composer.selection, isNotNull);
+
+      // Simulate the user copying a full markdown document
```
Comment (angelosilvestre) [inline super_editor_markdown/test/super_editor_markdown_pasting_test.dart:61]: Please end the comment with a `.`.

File: super_editor_markdown/test/super_editor_markdown_pasting_test.dart [hunk 7]
```diff
@@ -0,0 +1,424 @@
...
+      '''),
+      );
+
+      // Place the caret at the beginning of the document.
+      await tester.placeCaretInParagraph(document.nodes.first.id, 0);
+
+      // Ensure that the document has the caret.
+      expect(composer.selection, isNotNull);
+
+      // Simulate the user copying a markdown snippet
```
Comment (angelosilvestre) [inline super_editor_markdown/test/super_editor_markdown_pasting_test.dart:137]: Please end the comment with a `.`.

File: super_editor_markdown/test/super_editor_markdown_pasting_test.dart [hunk 8]
```diff
@@ -0,0 +1,424 @@
...
+      );
+    });
+
+    testWidgetsOnArbitraryDesktop("can paste in the middle of a document and merge both sides of text", (tester) async {
+      final (_, document, composer) = await _pumpSuperEditor(
+        tester,
+        deserializeMarkdownToDocument('''This is a paragraph that will split >< here and continue.'''),
+      );
+
+      // Place the caret between chevrons ">|<"
```
Comment (angelosilvestre) [inline super_editor_markdown/test/super_editor_markdown_pasting_test.dart:170]: Please end the comment with a `.`.

File: super_editor_markdown/test/super_editor_markdown_pasting_test.dart [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+      final lastParagraph = document.nodes.last as TextNode;
+      await tester.placeCaretInParagraph(lastParagraph.id, 37);
+
+      // Ensure that the document has the caret.
+      expect(composer.selection, isNotNull);
+
+      // Simulate the user copying a markdown snippet
```
Comment (angelosilvestre) [inline super_editor_markdown/test/super_editor_markdown_pasting_test.dart:177]: Please end the comment with a `.`.

File: super_editor_markdown/test/super_editor_markdown_pasting_test.dart [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+      );
+    });
+
+    testWidgetsOnArbitraryDesktop("can paste a text snippet within a single paragraph", (tester) async {
+      final (_, document, composer) = await _pumpSuperEditor(
+        tester,
+        deserializeMarkdownToDocument('''This is a paragraph that will add text >< here and continue.'''),
+      );
+
+      // Place the caret between chevrons ">|<"
```
Comment (angelosilvestre) [inline super_editor_markdown/test/super_editor_markdown_pasting_test.dart:205]: Please end the comment with a `.`.

File: super_editor_markdown/test/super_editor_markdown_pasting_test.dart [hunk 11]
```diff
@@ -0,0 +1,424 @@
...
+      final lastParagraph = document.nodes.last as TextNode;
+      await tester.placeCaretInParagraph(lastParagraph.id, 40);
+
+      // Ensure that the document has the caret.
+      expect(composer.selection, isNotNull);
+
+      // Simulate the user copying a markdown snippet
```
Comment (angelosilvestre) [inline super_editor_markdown/test/super_editor_markdown_pasting_test.dart:212]: Please end the comment with a `.`.

File: super_editor_markdown/test/super_editor_markdown_pasting_test.dart [hunk 12]
```diff
@@ -0,0 +1,424 @@
...
+
+    testWidgetsOnArbitraryDesktop("can paste an image in the middle of a paragraph", (tester) async {
+      final (_, document, composer) = await _pumpSuperEditor(
+        tester,
+        deserializeMarkdownToDocument(
+          '''This is a paragraph that will split here >< and show an image between paragraphs.''',
+        ),
+      );
+
+      // Place the caret between chevrons ">|<"
```
Comment (angelosilvestre) [inline super_editor_markdown/test/super_editor_markdown_pasting_test.dart:238]: Please end the comment with a `.`.

File: super_editor_markdown/test/super_editor_markdown_pasting_test.dart [hunk 13]
```diff
@@ -0,0 +1,424 @@
...
+      final lastParagraph = document.nodes.last as TextNode;
+      await tester.placeCaretInParagraph(lastParagraph.id, 42);
+
+      // Ensure that the document has the caret.
+      expect(composer.selection, isNotNull);
+
+      // Simulate the user copying a markdown snippet
```
Comment (angelosilvestre) [inline super_editor_markdown/test/super_editor_markdown_pasting_test.dart:245]: Please end the comment with a `.`.

File: super_editor_markdown/test/super_editor_markdown_pasting_test.dart [hunk 14]
```diff
@@ -0,0 +1,424 @@
...
+      // the starting and ending text of the pasted Markdown.
+      final documentMarkdown = serializeDocumentToMarkdown(document);
+
+      expect(
+        documentMarkdown,
+        '''This is a paragraph that will split here >
+
+![A Fake Test Image](https://flutter.dev/logo.png)
+< and show an image between paragraphs.''',
+      );
```
Comment (angelosilvestre) [inline super_editor_markdown/test/super_editor_markdown_pasting_test.dart:263]: We are not merging, right?

File: super_editor_markdown/test/super_editor_markdown_pasting_test.dart [hunk 15]
```diff
@@ -0,0 +1,424 @@
...
+
+      // Place the caret at the end of the existing document.
+      final lastParagraph = document.nodes.last as TextNode;
+      await tester.placeCaretInParagraph(lastParagraph.id, lastParagraph.endPosition.offset);
+
+      // Ensure that the document has the caret.
+      expect(composer.selection, isNotNull);
+
+      // Simulate the user copying a markdown snippet
```
Comment (angelosilvestre) [inline super_editor_markdown/test/super_editor_markdown_pasting_test.dart:283]: Please end the comment with a `.`.

File: super_editor_markdown/test/super_editor_markdown_pasting_test.dart [hunk 16]
```diff
@@ -0,0 +1,424 @@
...
```
Comment (angelosilvestre) [inline super_editor_markdown/test/super_editor_markdown_pasting_test.dart:328]: Please end the comment with a `.`.

**Review by angelosilvestre — APPROVED** [review]
LGTM with some comments.

File: super_editor/lib/src/default_editor/multi_node_editing.dart (hunk 1 above)
Comment (matthew-carroll) [reply super_editor/lib/src/default_editor/multi_node_editing.dart:30]: I guess I should probably add the comment to the constructor and reference the parameters instead of hte properties. The question of public vs private should be based on code usage rather than Dart Docs.

File: super_editor/lib/src/default_editor/multi_node_editing.dart (hunk 5 above)
Comment (matthew-carroll) [reply super_editor/lib/src/default_editor/multi_node_editing.dart:222]: I don't think so. There are two different perspectives on this action: the meaning going in, and the meaning going out. When choosing what information to pass in, the name `currentNodeWithSelectionId` makes it clear what the caller should provide. Also, `newNodeId` clearly distinguishes itself from the existing node with the selection. So those are generally helpful names.

However, when working with the response, regardless of which node is new or old, selected or not selected, what you want the method to produce is one paragraph split into two, in which case one is upstream and one is downstream.

File: super_editor_markdown/test/super_editor_markdown_pasting_test.dart (hunk 14 above)
Comment (matthew-carroll) [reply super_editor_markdown/test/super_editor_markdown_pasting_test.dart:263]: I'm not sure what you mean. Can you rephrase the question?

Comment (angelosilvestre) [reply super_editor_markdown/test/super_editor_markdown_pasting_test.dart:263]: The comment says we are merging the starting and ending text, but it doesn't seem we are doing that in this test.
